src/mds/SimpleLock.h

```diff
@@ -417,6 +417,7 @@ class SimpleLock {
     ceph_assert(state == LOCK_XLOCK || state == LOCK_XLOCKDONE ||
 	   state == LOCK_XLOCKSNAP || state == LOCK_LOCK_XLOCK ||
 	   state == LOCK_LOCK  || /* if we are a leader of a peer */
+	   state == LOCK_PREXLOCK ||
```

Comment:
@lxbsz From your comment on the tracker https://tracker.ceph.com/issues/44565#note-18, How does the lock state moved to LOCK_PREXLOCK on first decrement of x from 2. Logically it make sense to move to  LOCK_PREXLOCK on x = 0 ? I don't understand the lock state machine. Sorry if it's a naive question.

Reply:
No, it doesn't much matter with the `x` and the state still could be moved even it's not `0`.

This issue happened when one request tried to acquire the xlock for `ilink` locker:

Before moving the state to `LOCK_PREXLOCK` the state was `LOCK_LOCK_XLOCK`, which is a intermediate state before successfully setting it to `LOCK_XLOCK`. MDS will set the locker to a intermediate state if it couldn't move to the next target state directly, in this case it needed to revoke some caps from clients first.

From `mds/lock.cc` file we can see it's next state is `LOCK_PREXLOCK`:

```
[LOCK_LOCK_XLOCK]= { LOCK_PREXLOCK,false,LOCK_LOCK,0,   XCL, 0,   0,   0,   0,   XCL, 0,0,0,0 },
```
And just after the caps was successfully revoked and when continued moving the state to `LOCK_XLOCK`, the `eval_gather()` just set it to `LOCK_PREXLOCK`. And then later when the corresponding request was retried it should move the state to `LOCK_XLOCK`.

Reply:
Thank @lxbsz for the explaination.